### DISCUSSION TITLE
Master/Server

### DISCUSSION THREAD

**Discussion by jeserama** [eModbus/eModbus#113] [eModbus applications]
Hi guys first of all I've read the documentation and the work is really amazing. I have to deal with modbus protocol but I'm really not familiar with this protocol. 
So I have two ESP32 : the one should be the master and the other the slave. The communication is with GPIO_NUM_16 and GPIO_NUM_17.
But when I try to run the master ( using RTU06example) , I get some error : 
```
                          ␛[0m␛[1;33m[E] 148| main.cpp             [  40] handleError: Error response: E4 - Server ID mismatch
```
Note that I've already define  server_id as 1 .
The code for the server is the same as in the example RTUServer.
Thanks in advance

**Comment by Miq1** [comment]
This error message points to a different server ID found in the response than was sent with the request. This must not happen, as the requested server has to respond with its own server ID and nothing else.

**Reply by jeserama** [reply]
When I send the request with: MB.addRequest(Token++, server_id,READ_HOLD_REGISTER , 10, 123) where server_id  is 3 . and And in the server  MBserver.registerWorker(server_id, READ_HOLD_REGISTER, &FC03);.

So I don't really understande why I got this error

**Comment by bertmelis** [comment]
How did you connect the two ESP's?

**Reply by jeserama** [reply]
I use a crosswise to the ESP GPIO16 and GPIO17.

**Reply by bertmelis** [reply]
So you just connect the RX of esp 1 to TX of esp 2 and vice versa?

Can you enable the debug output? --> https://emodbus.github.io/logging

**Reply by Miq1** [reply]
...and for better insights you should show us both of your client and server sketches.

**Reply by jeserama** [reply]
Yes I just connect  RX of esp 1 to TX of esp 2 and vice versa!

This is my client code :
```cpp
// =================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================

// Example code to show the usage of the eModbus library. 
// Please refer to root/Readme.md for a full description.

// Includes: <Arduino.h> for Serial etc.
#include <Arduino.h>

// Include the header for the ModbusClient RTU style
#include "ModbusClientRTU.h"
#define LOCAL_LOG_LEVEL LOG_LEVEL_6
// For demonstration, use the LOG statements for output
#include "Logging.h"
#define server_id 0x0003 
// Create a ModbusRTU client instance
// In my case the RS485 module had auto halfduplex, so no second parameter with the DE/RE pin is required!
ModbusClientRTU MB(Serial2);

// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the message data and length of it, 
// plus a user-supplied token to identify the causing request
void handleData(ModbusMessage response, uint32_t token) 
{
  // Only print out result of the "real" example - not the request preparing the field
  if (token > 1111) {
    LOG_N("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
    HEXDUMP_N("Data", response.data(), response.size());
  }
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  LOG_E("Error response: %02X - %s\n", (int)me, (const char *)me);
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Set up Serial2 connected to Modbus RTU
// (Fill in your data here!)
  Serial2.begin(19200, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);

// Set up ModbusRTU client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task
  MB.begin();

// We will first set the register to a known state, read the register, 
// then write to it and finally read it again to verify the change

// Set defined conditions first - write 0x1234 to the register
// The Token value is used in handleData to avoid the output for this first preparation request!
  uint32_t Token = 1111;

  Error err = MB.addRequest(Token++, server_id,READ_HOLD_REGISTER , 10, 123);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// Create request by
// (Fill in your data here!)
// - token to match the response with the request.
// - server ID = 1
// - function code = 0x03 (read holding register)
// - address to read = word 10
// - data words to read = 1
//
// If something is missing or wrong with the call parameters, we will immediately get an error code 
// and the request will not be issued

// Read register
  err = MB.addRequest(Token++, server_id, READ_HOLD_REGISTER, 10, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// Now write 0xBEEF to it
  err = MB.addRequest(Token++, server_id, WRITE_HOLD_REGISTER, 10, 0xBEEF);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// Read it again to verify
  err = MB.addRequest(Token++, server_id, READ_HOLD_REGISTER, 10, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// Provoke an error just to show it
  err = MB.addRequest(Token++, 3, USER_DEFINED_44, 10, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// The output on the Serial Monitor will be (depending on your Modbus the data will be different):
//     __ OK __
//     [N] 163| main.cpp             [  29] handleData: Response: serverID=1, FC=3, Token=00000458, length=5:
//     [N] Data: @3FFB9808/5:
//       | 0000: 01 03 02 12 34                                    |....4           |
//     [N] 185| main.cpp             [  29] handleData: Response: serverID=1, FC=6, Token=00000459, length=6:
//     [N] Data: @3FFB9808/6:
//       | 0000: 01 06 00 0A BE EF                                 |......          |
//     [N] 209| main.cpp             [  29] handleData: Response: serverID=1, FC=3, Token=0000045A, length=5:
//     [N] Data: @3FFB9808/5:
//       | 0000: 01 03 02 BE EF                                    |.....           |
//     [E] 231| main.cpp             [  40] handleError: Error response: 01 - Illegal function code
}

// loop() - nothing done here today!
void loop() {
}
```

**Reply by jeserama** [reply]
And this is the server
```cpp
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================
// Includes: <Arduino.h> for Serial etc., WiFi.h for WiFi support
#include <Arduino.h>
#include "HardwareSerial.h"
#define LOCAL_LOG_LEVEL LOG_LEVEL_6
// Modbus server include
#include "ModbusServerRTU.h"
#include "Logging.h"
#define server_id 0x0003 
// Create a ModbusRTU server instance listening on Serial2 with 2000ms timeout
ModbusServerRTU MBserver(Serial2, 2000);

// FC03: worker do serve Modbus function code 0x03 (READ_HOLD_REGISTER)
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;           // requested register address
  uint16_t words;             // requested number of registers
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, words);

  // Address and words valid? We assume 10 registers here for demo
  if (address && words && (address + words) <= 10) {
    // Looks okay. Set up message with serverID, FC and length of data
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(words * 2));
    // Fill response with requested data
    for (uint16_t i = address; i < address + words; ++i) {
      response.add(i);
    }
  } else {
    // No, either address or words are outside the limits. Set up error response.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Init Serial2 connected to the RTU Modbus
// (Fill in your data here!)
  Serial2.begin(19200, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);

// Register served function code worker for server 1, FC 0x03
  MBserver.registerWorker(server_id, READ_HOLD_REGISTER, &FC03);

  

// Start ModbusRTU background task
  MBserver.start();
}

// loop() - nothing done here today!
void loop() {

  delay(10000);
}
```

**Reply by Miq1** [reply]
At a first glance the code looks okay - except the client will request function codes the server is not providing, but that may be on intent.

I would really be interested in the debug outputs of both client and server.

**Reply by Miq1** [reply]
Signore? Anything you can show us?

**Comment by brhmira** [comment]
Hello guys, the library looks really nice and easy to use. Unfortunately, I couldn't get it working yet. I am using the very same setup as @jeserama did:
- two esp32 (connected crosswise),
- HardwareSerial Serial2
- MBServer implemented like this:
``MBserver.registerWorker(server_id, READ_HOLD_REGISTER, &FC03);``
``MBserver.start();``
- MBClient implemented similarly to the example codes (just MB.onDataHandler, MB.onErrorHandler, MB.setTimeout, MB.begin calls in the setup function) and then I put MB.addRequest in a loop (with some interval, copy-paste from RTU04example)

**Unfortunately, I am getting only the E0 timeout error all the time.** The client is probably sending the request correctly, but the server never ever receives it. The callback `FC03` of the registered worker is never called.
I am quite sure that serverID and FC is matching on both sides.
I can send and recive messages without modbus protocol easily via Serial2. The Serial2.println("hello world); and Serial2.readUntil() are working as a charm on the same HW setup.
I am quite desperate now.

I have also tested it with two TTL to RS485 (without DE/RE pins) (similar to this: https://www.ebay.com/itm/TTL-To-RS485-Adapter-485-Serial-Port-UART-Level-Converter-Function-Module-3-3v-/372862798390),

I can post my code and debug output later. I was just curious whether the library should work even with the basic setup (two esp32 connected directly, crosswise). 

Thank you!

**Reply by Miq1** [reply]
> I can post my code and debug output later. I was just curious whether the library should work even with the basic setup (two esp32 connected directly, crosswise).

Please do so. It should be no problem at all connecting two ESP32s the way you wrote. I suppose the baud rates etc. are set to matching values on both ESP32s?

**Reply by brhmira** [reply]
@Miq1 Thank you for the quick response. It helped me! 
I didn't mention in my original comment that I was trying to integrate the eModbus library into a bigger project. The need to post my code forced me to remove all other code (not related to Modbus communication). And then the communication started to work!
So the issue is not in the library, wiring, or Modbus code.

I thought that Modbus communication is performed in a separate thread, therefore it is not interfering with the rest of the code (reading sensors, wifi, delays, etc...). But apparently, it somehow interferes. I hope I can figure it out on my own.

**Reply by Miq1** [reply]
> I thought that Modbus communication is performed in a separate thread, therefore it is not interfering with the rest of the code (reading sensors, wifi, delays, etc...). But apparently, it somehow interferes. I hope I can figure it out on my own.

It is, in fact. The RTU timing is very critical, so the task priority needs to be as close to real time as possible to properly work. If in the middle of a RTU receive operation the task is switched, this may damage the message. I tried to keep the spots for task changes to a minimum, but still there are some.

**Reply by bertmelis** [reply]
You can try to have the modbus thread running on a different core: https://emodbus.github.io/modbusserver-rtu-api#bool-start-bool-startint-coreid-and-bool-startint-coreid-uint32_t-interval
